### PR TITLE
Fix AWS SSO access

### PR DESCRIPTION
**Pull request by mfittko**
We were having issues when using the gem with AWS SSO credentials (profiles without fixed credentials). This patch fixes the issue for us.

### CODE REVIEW

File: lib/elastic_whenever/option.rb [hunk 1]
```diff
@@ -133,10 +133,11 @@ def initialize(args)
                      elsif access_key && secret_key
                        Aws::Credentials.new(access_key, secret_key)
                      end
+      @credentials = nil unless @credentials&.set?
     end
 
     def aws_config
-      @aws_config ||= { credentials: credentials, region: region }.delete_if { |_k, v| v.nil? }
+      @aws_config ||= { credentials: credentials, region: region, profile: profile }.delete_if { |_k, v| v.nil? }
```
Comment (wata727) [inline lib/elastic_whenever/option.rb:140]: Can you explain why this change is necessary? It looks like you need to specify the profile name, but my understanding is that passing `profile_name` in `Aws::SharedCredentials` should be an alternative to this.
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/#Configuration_Options
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SharedCredentials.html

**Review by wata727 — COMMENTED** [review]
CI is failing and need to fix specs.

File: lib/elastic_whenever/option.rb (hunk 1 above)
Comment (mfittko) [reply lib/elastic_whenever/option.rb:140]: with AWS SSO, I was not able to get the shared credentials to work as credentials was always `nil`. It's not necessary to set credentials at all in the case you are logged into SSO. The SDK should handle this as long as you have a profile set.

**Comment by mfittko** [comment]
Sorry, but I did not set up the specs on my fork. I can fix them locally.

**Comment by mfittko** [comment]
@wata727 specs are green and I made sure that the profile isn't passed if there are credentials set

File: lib/elastic_whenever/option.rb (hunk 1 above)
Comment (wata727) [reply lib/elastic_whenever/option.rb]: Hmm, I did a little research and couldn't figure out why a profile need to be configured differently for SSO than others.
Is it not necessary to set `Aws::SSOCredentials` for SSO?
https://docs.aws.amazon.com/sdk-for-ruby/v3/developer-guide/credentials.html
https://docs.aws.amazon.com/sdkref/latest/guide/feature-sso-credentials.html
https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/sso.html
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SSOCredentials.html

Comment (mfittko) [reply lib/elastic_whenever/option.rb]: well, same with instance credentials, there's no need for explicit credentials when using SSO afaik

Comment (mfittko) [reply lib/elastic_whenever/option.rb]: at least this change works for me. we also successfully run elastic_whenever on AWS codebuild, where also no explicit credentials are present.

Comment (mfittko) [reply lib/elastic_whenever/option.rb]: if you prefer to pass explicit credentials, go for it, but IMO it's a no-op

Comment (mfittko) [reply lib/elastic_whenever/option.rb]: I could actually go around that change if I'd set `AWS_PROFILE`, I think! so should I revert the line?

File: lib/elastic_whenever/option.rb [hunk 2]
```diff
@@ -133,10 +133,11 @@ def initialize(args)
                      elsif access_key && secret_key
                        Aws::Credentials.new(access_key, secret_key)
                      end
+      @credentials = nil unless @credentials&.set?
```
Comment (mfittko) [inline lib/elastic_whenever/option.rb:136]: just a little explainer: when using SSO credentials, the `Aws::SharedCredentials` instance's `credentials` will not be set! that's why I chose to `nil` the `@credentials` again in that case.

File: lib/elastic_whenever/option.rb [hunk 3]
```diff
@@ -133,10 +133,11 @@ def initialize(args)
...
     end
 
     def aws_config
-      @aws_config ||= { credentials: credentials, region: region }.delete_if { |_k, v| v.nil? }
+      @aws_config ||= { credentials: credentials, region: region, profile: @credentials ? nil : profile }.delete_if { |_k, v| v.nil? }
```
Comment (mfittko) [inline lib/elastic_whenever/option.rb:140]: if the `@credentials` are nil due to whatever reason, we should pass in the profile in order to actually support the `--profile` option as it would otherwise just be ignored. A workaround could be to set the `AWS_PROFILE` env var in that case, but why pass in the `region`, too then?

Comment (mfittko) [reply lib/elastic_whenever/option.rb:140]: using SSO, it is not required to set credentials for an AWS SDK client at all, but you can switch profiles by passing in the profile as can be seen in this example:
```
irb(main):001> require 'aws-sdk-sts'
=> true
irb(main):002> sts_client = Aws::STS::Client.new
=> #<Aws::STS::Client>
irb(main):003> response = sts_client.get_caller_identity
=>
#<struct Aws::STS::Types::GetCallerIdentityResponse
...
irb(main):004> puts "Account: #{response.account}"
Account: 493373280XXX
=> nil
irb(main):005> sts_client = Aws::STS::Client.new profile: 'sofatutor-testing'
=> #<Aws::STS::Client>
irb(main):006> response = sts_client.get_caller_identity
=>
#<struct Aws::STS::Types::GetCallerIdentityResponse
...
irb(main):007> puts "Account: #{response.account}"
Account: 800587838XXX

File: lib/elastic_whenever/option.rb (hunk 1 above)
Comment (wata727) [reply lib/elastic_whenever/option.rb]: Ah, I understand. SSO uses `profile` regardless of shared credentials. I misunderstood that profile is a concept only for shared credentials.
Then this change makes sense. It's okay as is.

File: lib/elastic_whenever/option.rb (hunk 2 above)
Comment (wata727) [reply lib/elastic_whenever/option.rb:136]: Considering that `profile` is not just for shared credentials, I think it is probably not appropriate to return `Aws::SharedCredentials` when the `profile` exists.
It might make more sense if including the `profile` in `aws_config` means you don't have to explicitly return `SharedCredentials` here.

e.g.

```ruby
@credentials = if access_key && secret_key
                 Aws::Credentials.new(access_key, secret_key)
               end
```

File: lib/elastic_whenever/option.rb (hunk 3 above)
Comment (wata727) [reply lib/elastic_whenever/option.rb:140]: Thank you for your explanation. With this explanation, I understand that `profile` can be referenced other than shared credentials.

File: lib/elastic_whenever/option.rb (hunk 2 above)
Comment (wata727) [reply lib/elastic_whenever/option.rb:136]: Yeah, I agree that this approach is valid. However, since there is no need to explicitly return `SharedCredentials` in the first place, I'm concerned that when looking at it later, it may be difficult to understand why `SharedCredentials` is returned here. At least I couldn't understand why this line was needed without explanation.

It is possible to merge this change, but I will probably release it after fixing it as above.

Comment (mfittko) [reply lib/elastic_whenever/option.rb:136]: Okay question is how you would find out that SharedCredentials aren't needed in the first place without adding a lot more code. If you feel like adding more code for solving the same problem, it's up to you. Happy if you just take it up then. Meanwhile we can keep using our fork then.

Comment (mfittko) [reply lib/elastic_whenever/option.rb:136]: Maybe I got you wrong, so you propose to just remove the shared credentials completely? I can try and patch up the tests.

Comment (mfittko) [reply lib/elastic_whenever/option.rb]: I applied the change, I think it's valid to remove the SharedCredentials completely since AWS sdk should handle the credentials if none are passed via a key pair.

Comment (mfittko) [reply lib/elastic_whenever/option.rb]: Not to mention that passing in an AWS key pair via CLI options is a security risk as credentials will pretty likely end up in history or logs. But that's not my concern here :)

File: lib/elastic_whenever/option.rb [hunk 4]
```diff
@@ -128,15 +128,13 @@ def initialize(args)
...
-                       Aws::SharedCredentials.new(profile_name: profile)
-                     elsif access_key && secret_key
+      @credentials = if access_key && secret_key
                        Aws::Credentials.new(access_key, secret_key)
                      end
     end
 
     def aws_config
-      @aws_config ||= { credentials: credentials, region: region }.delete_if { |_k, v| v.nil? }
+      @aws_config ||= { credentials: credentials, region: region, profile: @credentials ? nil : profile }.delete_if { |_k, v| v.nil? }
```
Comment (wata727) [inline lib/elastic_whenever/option.rb:137]: Do we need to check for `@credentials` here? I'm thinking it might be a good idea to simply pass the `profile`.

```suggestion
      @aws_config ||= { credentials: credentials, region: region, profile: profile }.delete_if { |_k, v| v.nil? }
```

Comment (mfittko) [reply lib/elastic_whenever/option.rb:137]: I think passing it would be redundant since there's no need for looking up the profile and I wouldn't even know what would happen when someone would pass profile _and_ credentials 🤔

Comment (wata727) [reply lib/elastic_whenever/option.rb:137]: I agree that there is no valid case for passing a profile with static credentials, but I also don't think there is a need to intentionally ignore a profile passed with `--profile`.
I believe that it is better to reflect the flags specified by users as is, to keep the implementation simple and intuitive.

**Review by wata727 — APPROVED** [review]
Thanks!
